parsl/serialize/facade.py

```diff
@@ -0,0 +1,174 @@
+        b_kwargs = self.serialize(kwargs, buffer_threshold=buffer_threshold)
+        packed_buffer = self.pack_buffers([b_func, b_args, b_kwargs])
+        return packed_buffer
+
+    def unpack_apply_message(self, packed_buffer, user_ns=None, copy=False):
+        """ Unpack and deserialize function and parameters
+
+        """
+        return [self.deserialize(buf) for buf in self.unpack_buffers(packed_buffer)]
+
+    def serialize(self, obj, buffer_threshold=1e6):
+        """ Try available serialization methods one at a time
+
+        Individual serialization methods might raise a TypeError (eg. if objects are non serializable)
+        This method will raise the exception from the last method that was tried, if all methods fail.
+        """
+        serialized = None
+        serialized_flag = False
+        last_exception = None
+        if callable(obj):
+            for method in self.methods_for_code.values():
+                try:
+                    serialized = method.serialize(obj)
+                    # We attempt a deserialization to make sure both work.
+                    method.deserialize(serialized)
+                except Exception as e:
+                    logger.exception(f"Serialization method: {method} did not work")
+                    last_exception = e
+                    continue
+                else:
+                    serialized_flag = True
+                    break
+        else:
+            for method in self.methods_for_data.values():
+                try:
+                    serialized = method.serialize(obj)
+                except Exception as e:
+                    logger.exception(f"Serialization method {method} did not work")
+                    last_exception = e
+                    continue
+                else:
+                    serialized_flag = True
+                    break
+
+        if serialized_flag is False:
+            # TODO : Replace with a SerializationError
+            raise last_exception
+
+        if len(serialized) > buffer_threshold:
+            raise TypeError(f"Serialized object is too large and exceeds buffer threshold of {buffer_threshold} bytes")
+        return serialized
+
+    def deserialize(self, payload):
+        """
+        Parameters
+        ----------
+        payload : str
+           Payload object to be deserialized
+
+        """
+        header = payload[0:self.header_size]
+        if header in self.methods_for_code:
+            result = self.methods_for_code[header].deserialize(payload)
+        elif header in self.methods_for_data:
+            result = self.methods_for_data[header].deserialize(payload)
+        else:
+            raise TypeError("Invalid header: {} in data payload. Buffer is either corrupt or not created by ParslSerializer".format(header))
+
+        return result
+
+    def pack_buffers(self, buffers):
+        """
+        Parameters
+        ----------
+        buffers : list of \n terminated strings
```

Comment:
this \n requirement is weird. It looks like the individual strings can contain \n anywhere in them because the \n isn't used to signify anything and packing/unpacking is done by transmitting the string length.
I don't think this really is a requirement?